### ISSUE TITLE
OpaquePeerId is not displayed anymore

### ISSUE THREAD

**Issue by gdethier** [polkadot-js/apps#8113]
* **I'm submitting a ...**

  - [x] Bug report
  - [ ] Feature request
  - [ ] Support request
  - [ ] Other

* **What is the current behavior and expected behavior?**

In a chain with `nodeAuthorization` pallet installed, when looking at the well-known nodes, an empty list appears:

![empty well-known nodes list](https://user-images.githubusercontent.com/883394/189315075-a893f389-10bd-4dd9-8f6d-82c44c8ff2ba.png)

Previously, the peer ID of each well known node was displayed as a hex string (but without the "0x" prefix). In the above image, one can see that there is an empty space for the single well-known node, but the peer ID is not displayed at all. Note that we are using the `OpaquePeerId` type in another of our pallets, and the problem is the same. Below an example taken from a chain with 4 well-known nodes, there is one empty line per node:

![empty well-known nodes list](https://user-images.githubusercontent.com/883394/189317813-ef31310c-adf0-48ac-b0fc-df4ad0d8b319.png)

It would be nice to, at least, come back to the initial behavior (see the hex representation of the peer ID). Adding the "0x" prefix to the hex string would be a plus.

* **What is the motivation for changing the behavior?**

Fix a regression. Furthermore, when managing such a chain, it is necessary to be able to visualize the well-known nodes.

* **Please tell us about your environment:**

- Chain: [logion Node v4.0.0](https://github.com/logion-network/logion-node/tree/645cab9566871cb27f72453b9a3c58f98217677c)
- api v9.3.3
- apps v0.121.2-42-x

In order to take the first screenshot above, the logion node (follow above link to get the exact version) was executed using `./target/release/logion-node --dev` so the chain is running in `Development` mode and using the `dev` chain spec which defines one single authorized node and validator.

  - Version:
  - Environment:

    - [ ] Node.js
    - [x] Browser
    - [ ] Other (limited support for other environments)

  - Language:

    - [ ] JavaScript
    - [ ] TypeScript (include tsc --version)
    - [ ] Other

**Comment by jacogr** [comment]
Is there a running chain with this issue?

**Comment by gdethier** [comment]
Yes, see here: https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Ftest-rpc01.logion.network#/chainstate

**Comment by jacogr** [comment]
On the running chain you linked, the encoding of the Id is not correct.

So basically, your state returns the following -

```js
0x // hex prefix
10 // length, 4 items
// items
980024080112201ce5f00ef6e89374afb625f1ae4c1546d31234e87e3c3f51a62b91dd6bfa57df
98002408011220876a7b4984f98006dc8d666e28b60de307309835d775e7755cc770328cdacf2e
98002408011220c81bc1d7057a1511eb9496f056f6f53cdfe0e14c8bd5ffca47c70a8d76c1326d
98002408011220dacde7714d8551f674b8bb4b54239383c76a2b286fa436e93b2b7eb226bf4de7
```

In this case, Opaque is meant to have 38 bytes and then the inner is meant to have 0 bytes... It should be double-encoded, e.g. as part of a heartbeat, it looks like -

```
9c // length
98 // length
// bytes
00240801122017f7786847b1788be1734acc5230afdb7abde2d1818730fc6e0e4eb6a69e1c86
```

In that case, Opaque is 39 bytes and the inner Bytes has 37 bytes. (So it is `<compact length><compact length><data>`). In your case the second compact-length is 0, so it returns an empty string in display, e.g.

```
98 // length
00 // zero length
// bytes would be ignored since the length is set to 0 above
24080112201ce5f00ef6e89374afb625f1ae4c1546d31234e87e3c3f51a62b91dd6bfa57df
```

**Comment by jacogr** [comment]
Follow-up, looked at your raw state for test in the repo, it is -

```
 "0x6b219fca3e055e58d26bd69d8d9f7edb267722115100ff8c65fc1ec2ac0e893f": "0x0c980024080112201ce5f00ef6e89374afb625f1ae4c1546d31234e87e3c3f51a62b91dd6bfa57df98002408011220876a7b4984f98006dc8d666e28b60de307309835d775e7755cc770328cdacf2e98002408011220dacde7714d8551f674b8bb4b54239383c76a2b286fa436e93b2b7eb226bf4de7",
```

It should be -

```
"0x6b219fca3e055e58d26bd69d8d9f7edb267722115100ff8c65fc1ec2ac0e893f": "0x0c9c980024080112201ce5f00ef6e89374afb625f1ae4c1546d31234e87e3c3f51a62b91dd6bfa57df9c98002408011220876a7b4984f98006dc8d666e28b60de307309835d775e7755cc770328cdacf2e9c98002408011220dacde7714d8551f674b8bb4b54239383c76a2b286fa436e93b2b7eb226bf4de7",
```

An extra `9c` before the `980024...` occurrences.

**Comment by gdethier** [comment]
@jacogr Thanks for the explanation! If I understand well, the problem must be fixed at the node side?

**Comment by jacogr** [comment]
For you test node, update the raw spec files. For the running chains, you can update the state on-chain via system.setStorage.

**Comment by gdethier** [comment]
OK thanks. However, the state gets corrupted each time I insert new data (e.g. when I add a new well-known node) so I guess there is something going wrong upon actual write.

**Comment by jacogr** [comment]
Since your encoding in the state is basically `Vec<u8>` as opposed to `Opaque<Vec<u8>>`, that could be the cause. As soon as you try and change something, it goes wrong.

When you use `system.setStorage` you can actually write the full data. Try it on dev re-writing it and then try to add a new node, it should not get corrupted again.

**Comment by gdethier** [comment]
@jacogr Following your instructions, I am indeed able to fix the state and have the Peer IDs properly displayed. When adding a new well known node using polkadot.js/apps, I have to prefix the peer ID bytes with `98` so that an `Opaque<Vec<u8>>` is sent instead of a `Vec<u8>`.

However, there is still something that I do not get. You mentioned above that the raw state in the chain spec is already wrong (the outer length is missing). However, I am generating that state directly using the node executable with the `build-spec` command. The `nodeAuthorization` pallet is configured in the genesis as explained in the [tutorial on permissioned networks](https://docs.substrate.io/tutorials/get-started/permissioned-network/), see [here](https://github.com/logion-network/logion-node/blob/main/node/src/chain_spec.rs#L219). Does that mean that the `nodeAuthorization` pallet is not properly encoding the state?

**Comment by gdethier** [comment]
I think I found something: there is a clash between `imOnline`'s `OpaquePeerId` and `nodeAuthorization`'s one. The `OpaquePeerId` of `imOnline` is indeed an `Opaque<Vec<u8>>` but `nodeAuthorization`'s one is not. If I override `OpaquePeerId` and force it to `Vec<u8>`, everything works as expected.

**Comment by jacogr** [comment]
Yes, indeed if you are using a different type, best is to just override it.

**Comment by gdethier** [comment]
Overriding the type directly in polkadot.js works but I am not able to generate properly the TypeScript definitions. Despite the fact that `OpaquePeerId` is overridden in my `definitions.ts` file, the `augment-*.ts` files still import `imOnline`'s type (i.e. I keep seeing `import type { OpaquePeerId } from '@polkadot/types/interfaces/imOnline';` import statement).

**Comment by gdethier** [comment]
As a workaround, I fixed the `augment-*.ts` files manually.

We can probably consider this issue as closed. I guess I should open a new issue for the code generation issue?

**Comment by jacogr** [comment]
I've looked at this again. And specifically the implemented types. I'm honestly not sure how, in the case of ImOnline, the stuff ends up being double-encoded - it is possibly due to the fact that it comes from libp2p which is already encoded. 

So actually in this case I don't think the API implementation is correct, although it is correct for the ImOnline case, but it seems to "work around" the double encoding visible there. It should be a simple `Vec<u8>` from my casual reading, but in this case it has to treat it as an `Opaque<Vec<u8>>`.

But yes, you are right, should be moved to the API repo and dissected for -

- generation (possibly correct if the API indeed does the next point correct for both cases)
- need to re-look at why the ImOnline output doesn't seem to get to what the type definitions would show

TL;DR This seems to be caused by an ImOnline oditity which the API encoding follows

**Comment by gdethier** [comment]
@jacogr I create the issue for the generation in API project: https://github.com/polkadot-js/api/issues/5219

Regarding the imOnline double-encoding part, I think my understanding is too limited to write a useful issue description, sorry.
